Userland/Libraries/LibELF/DynamicLinker.cpp

```diff
@@ -393,17 +394,25 @@ static Result<void, DlErrorMessage> link_main_library(DeprecatedString const& pa
         VERIFY(!result.is_error());
         auto& object = result.value();
 
+        if (loader.filepath().ends_with("/libc.so"sv)) {
+            initialize_libc(*object);
+        }
+
         if (loader.filepath().ends_with("/libsystem.so"sv)) {
             VERIFY(!loader.text_segments().is_empty());
             for (auto const& segment : loader.text_segments()) {
-                if (syscall(SC_msyscall, segment.address().get())) {
+                auto flags = static_cast<int>(VirtualMemoryRangeFlags::SyscallCode) | static_cast<int>(VirtualMemoryRangeFlags::Immutable);
+                if (syscall(SC_annotate_mapping, segment.address().get(), flags)) {
+                    VERIFY_NOT_REACHED();
+                }
+            }
+        } else {
+            for (auto const& segment : loader.text_segments()) {
+                auto flags = static_cast<int>(VirtualMemoryRangeFlags::Immutable);
+                if (syscall(SC_annotate_mapping, segment.address().get(), flags)) {
```

Comment:
Hmm this all looks correct but it's missing one case: who sets the mappings to immutable on dlopen?

Reply:
It seems like it does call the `link_main_library` here?
https://github.com/SerenityOS/serenity/blob/2693745336ddc6c7a790ecf8a760aa74651ac055/Userland/Libraries/LibELF/DynamicLinker.cpp#L497-L499

Otherwise please point me to where I need to add a call to this syscall.

Reply:
Per the discussion we had now in the Discord server, we could mark this as resolved.